src/carcockpit/gltf_viewer_widget.cpp

```diff
@@ -0,0 +1,218 @@
+	sc_renderer->set_scene(demoscene);
+
+	camrip = std::make_shared<ruis::render::camera>();
+	sc_renderer->set_external_camera(camrip);
+	sc_renderer->set_scene_scaling_factor(this->params.scaling_factor);
+	sc_renderer->set_environment_cube(this->params.environment_cube);
+}
+
+void gltf_viewer_widget::update(uint32_t dt)
+{
+	demoscene->update(dt);
+
+	this->fps_sec_counter += dt;
+	this->time += dt;
+	[[maybe_unused]] float ft = static_cast<float>(this->time) / std::milli::den;
+	float fdt = static_cast<float>(dt) / std::milli::den;
+	++this->fps;
+
+	float xx = 3 * cosf(ft / 2);
+	float zz = 3 * sinf(ft / 2);
+
+	auto light = demoscene->get_primary_light();
+	if (light) {
+		light->pos = {xx, 3, zz, 1};
+	}
+
+	if (this->fps_sec_counter >= std::milli::den) {
+		// std::cout << "fps = " << std::dec << fps << std::endl;
+		this->fps_sec_counter = 0;
+		this->fps = 0;
+	}
+
+	camera_position += (camera_attractor - camera_position) * fdt / camera_transition_duration;
+	ruis::vec3 remains = camera_attractor - camera_position;
+	ruis::real l2 = remains.x() * remains.x() + remains.y() * remains.y() + remains.z() * remains.z();
+	const ruis::real threshold = 0.00001;
+	if (l2 < threshold) {
+		camera_position = camera_attractor;
+	}
+
+	this->clear_cache();
+}
+
+void gltf_viewer_widget::toggle_camera(bool toggle)
+{
+	if (toggle)
+		camera_attractor = camera_position_top;
+	else
+		camera_attractor = camera_position_front;
+}
+
+void gltf_viewer_widget::set_normal_mapping(bool toggle)
+{
+	application::inst().shader_adv_v.set_normal_mapping(toggle);
+}
+
+constexpr float snap_speed = 1.07; // 1 is zero speed
+
+bool gltf_viewer_widget::on_mouse_button(const ruis::mouse_button_event& e)
+{
+	if (e.button == ruis::mouse_button::wheel_up) {
+		camera_attractor -= this->params.camera_target;
+		camera_attractor /= snap_speed;
+		camera_attractor += this->params.camera_target;
+
+		if (!this->params.smooth_navigation_zoom) {
+			camera_position -= this->params.camera_target;
+			camera_position /= snap_speed;
+			camera_position += this->params.camera_target;
+		}
+
+	} else if (e.button == ruis::mouse_button::wheel_down) {
+		camera_attractor -= this->params.camera_target;
+		camera_attractor *= snap_speed;
+		camera_attractor += this->params.camera_target;
+
+		if (!this->params.smooth_navigation_zoom) {
+			camera_position -= this->params.camera_target;
+			camera_position *= snap_speed;
+			camera_position += this->params.camera_target;
+		}
+
+	} else if (e.button == ruis::mouse_button::left) {
+		mouse_orbit = e.is_down;
+		if (e.is_down) {
+			mouse_changeview_start = e.pos;
+			camera_changeview_start = camera_position;
+		}
+	}
+	return true;
+}
+
+bool gltf_viewer_widget::on_mouse_move(const ruis::mouse_move_event& e)
+{
+	constexpr float mouse_orbit_speed_multiplier = 2.0f;
+
+	if (mouse_orbit) {
+		ruis::vec2 diff = e.pos - mouse_changeview_start;
+
+		ruis::vec3 cam2go = camera_changeview_start - this->params.camera_target;
```

Comment:
choose better variable name. `cam2go` is not descriptive, sounds like GoPro camera :D

Reply:
renamed to cam_pos_relative